crypto/ring-signature/src/ring_signature/mlsag_sign.rs

```diff
@@ -0,0 +1,343 @@
+}
+
+impl<'a> MlsagSignParams<'a> {
+    /// Sign a ring of input addresses and amount commitments using a modified
+    /// MLSAG
+    ///
+    /// Returns the signed key_image and c_zero scalar
+    pub fn sign(
+        &self,
+        ring: impl Ring,
+        // Note: this `mut rng` can just be `rng` if this is merged upstream:
+        // https://github.com/dalek-cryptography/curve25519-dalek/pull/394
+        rng: impl CryptoRngCore,
+        responses: &mut [CurveScalar],
+    ) -> Result<(KeyImage, CurveScalar), Error> {
+        let ring_size = ring.size();
+
+        if self.real_index >= ring_size {
+            return Err(Error::IndexOutOfBounds);
+        }
+
+        // Check buffer lengths are correct
+
+        // `responses` must contain `2 * ring_size` elements.
+        if responses.len() != 2 * ring_size {
+            return Err(Error::LengthMismatch(2 * ring_size, responses.len()));
+        }
+
+        // Ring must decompress.
+        ring.check()?;
+
+        // Setup signing context
+        let mut sign_ctx = MlsagSignCtx::init(self, rng, responses)?;
+
+        // Iterate around the ring, starting at real_index.
+        // NOTE: THIS REORDERING IS CRITICAL FOR PIECEWISE COMPUTATION
+        for n in 0..ring_size {
+            let i = (self.real_index + n) % ring_size;
+            let tx_out = &ring.index(i)?;
+
+            sign_ctx.update(self, i, tx_out)?;
+        }
+
+        // "Close the loop" by computing responses for the real index.
+        let (key_image, c_zero) = sign_ctx.finalise(self)?;
+
+        Ok((key_image, c_zero))
+    }
+}
+
+/// MLSAG signing context, supports incremental computation for ring signing.
+///
+/// WARNING: MISUSE OF THIS API MAY GENERATE INVALID RINGS,
+/// SEE [`MlsagSign`] FOR A PREFERRED HIGHER LEVEL ABSTRACTION
+#[derive(Debug, Zeroize)]
+pub struct MlsagSignCtx<R: AsMut<[CurveScalar]>> {
+    alpha_0: Scalar,
+    alpha_1: Scalar,
+    G: RistrettoPoint,
+    I: RistrettoPoint,
+
+    /// Cache real input for balance checking
+    #[zeroize(skip)]
+    real_input: Option<(RistrettoPublic, Commitment)>,
+
+    /// Number of rings computed
+    ring_count: usize,
+
+    responses: R,
+
+    real_challenge: Option<Scalar>,
+    last_challenge: Option<Scalar>,
+    zeroth_challenge: Option<Scalar>,
+
+    #[zeroize(skip)]
+    key_image: KeyImage,
+
+    #[zeroize(skip)]
+    output_commitment: Commitment,
+
+    complete: bool,
+}
+
+#[allow(dead_code)]
+impl<R: AsRef<[CurveScalar]> + AsMut<[CurveScalar]>> MlsagSignCtx<R> {
+    /// Initialise signing state, including `challenges` and `responses` working
+    /// buffers
+    pub fn init<'a>(
+        params: &'a MlsagSignParams<'a>,
+        mut rng: impl CryptoRngCore,
+        mut responses: R,
+    ) -> Result<Self, Error> {
+        let G = B_BLINDING;
+        debug_assert!(
+            params.generator.B_blinding == G,
+            "basepoint for blindings mismatch"
+        );
```

Comment:
Is this because of the multi asset types now?

Reply:
err, i just moved that from mlsag.rs:135?